### PR TITLE
[X86] Add support for MS inp functions.

### PR DESCRIPTION
**Pull request by MalaySanghi**
support _inp, _inpw, _inpd.
These functions were removed from the Windows runtime library, but aare still supported for kernel mode development.

### CODE REVIEW

**Comment by MalaySanghi** [comment]
This could probably have been a part of https://github.com/llvm/llvm-project/pull/93774 . If that's preferred, I'll add this change as a new commit to that PR.

**Comment by MalaySanghi** [comment]
tag @phoebewang @FreddyLeaf @RKSimon for review

File: clang/test/CodeGen/X86/ms-x86-intrinsics.c [hunk 1]
```diff
@@ -63,6 +63,82 @@ unsigned __int64 test__emulu(unsigned int a, unsigned int b) {
...
+// CHECK-I386-SAME: i16 noundef zeroext [[PORT:%.*]])
+// CHECK-I386-NEXT:  entry:
+// CHECK-I386-NEXT:    [[TMP0:%.*]] = tail call i32 asm sideeffect "inb ${1:b}, ${0:b}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 [[PORT]])
+// CHECK-I386-NEXT:    ret i32 [[TMP0]]
+//
+// CHECK-X64-LABEL: define dso_local i32 @test_inp(
+// CHECK-X64-SAME: i16 noundef [[PORT:%.*]])
+// CHECK-X64-NEXT:  entry:
+// CHECK-X64-NEXT:    [[TMP0:%.*]] = tail call i32 asm sideeffect "inb ${1:b}, ${0:b}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 [[PORT]])
+// CHECK-X64-NEXT:    ret i32 [[TMP0]]
```
Comment (phoebewang) [inline clang/test/CodeGen/X86/ms-x86-intrinsics.c:80]: These can be merged as
```
// CHECK-LABEL: define dso_local i32 @test_inp(
// CHECK-SAME: i16 noundef zeroext [[PORT:%.*]])
// CHECK-NEXT:  entry:
// CHECK-NEXT:    [[TMP0:%.*]] = tail call i32 asm sideeffect "inb ${1:b}, ${0:b}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 [[PORT]])
// CHECK-NEXT:    ret i32 [[TMP0]]
```

The same to others.

File: clang/test/CodeGen/X86/ms-x86-intrinsics.c [hunk 2]
```diff
@@ -63,6 +63,82 @@ unsigned __int64 test__emulu(unsigned int a, unsigned int b) {
 // CHECK: [[RES:%[0-9]+]] = mul nuw i64 [[Y]], [[X]]
 // CHECK: ret i64 [[RES]]
 
+
+int test_inp(unsigned short port) {
+  return _inp(port);
+}
+// CHECK-I386-LABEL: define dso_local i32 @test_inp(
+// CHECK-I386-SAME: i16 noundef zeroext [[PORT:%.*]])
```
Comment (phoebewang) [inline clang/test/CodeGen/X86/ms-x86-intrinsics.c:71]: These can be put in one line now.

Comment (MalaySanghi) [reply clang/test/CodeGen/X86/ms-x86-intrinsics.c:71]: No, I had tried that. A variable (PORT in this case) is not allowed in CHECK-LABEL

File: clang/test/CodeGen/X86/ms-x86-intrinsics.c (hunk 1 above)
Comment (MalaySanghi) [reply clang/test/CodeGen/X86/ms-x86-intrinsics.c:80]: Note that the 32b case has a zeroext but the 64b case doesn't

Comment (phoebewang) [reply clang/test/CodeGen/X86/ms-x86-intrinsics.c:80]: I think we don't care about zeroext or not here, you can simple check
`CHECK-LABEL: define dso_local i32 @test_inp(i16 noundef`
and omit the rest. Then you can use CHECK now.

File: clang/lib/Headers/intrin.h [hunk 3]
```diff
@@ -329,6 +329,28 @@ static __inline__ void __DEFAULT_FN_ATTRS __stosq(unsigned __int64 *__dst,
...
+
+static inline int _inp(unsigned short port) {
+  int ret;
+  __asm__ volatile("inb %b1, %b0" : "=a"(ret) : "Nd"(port));
+  return ret;
+}
+
+static inline unsigned short _inpw(unsigned short port) {
+  unsigned short ret;
+  __asm__ volatile("inb %w1, %b0" : "=a"(ret) : "Nd"(port));
```
Comment (phoebewang) [inline clang/lib/Headers/intrin.h:341]: You need `inw` here.

File: clang/lib/Headers/intrin.h [hunk 4]
```diff
@@ -329,6 +329,28 @@ static __inline__ void __DEFAULT_FN_ATTRS __stosq(unsigned __int64 *__dst,
...
+  return ret;
+}
+
+static inline unsigned long _inpd(unsigned short port) {
+  unsigned long ret;
+  __asm__ volatile("inb %k1, %b0" : "=a"(ret) : "Nd"(port));
```
Comment (phoebewang) [inline clang/lib/Headers/intrin.h:347]: `inl`

File: clang/test/CodeGen/X86/ms-x86-intrinsics.c [hunk 5]
```diff
@@ -63,6 +63,47 @@ unsigned __int64 test__emulu(unsigned int a, unsigned int b) {
...
+// CHECK-SAME:  [[PORT:%.*]])
+// CHECK:       [[TMP0:%.*]] = tail call i16 asm sideeffect "inw ${1:w}, ${0:b}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 [[PORT]])
+// CHECK-NEXT:  ret i16 [[TMP0]]
+
+unsigned long test_inpd(unsigned short port) {
+  return _inpd(port);
+}
+// CHECK-LABEL: i32 @test_inpd(i16 noundef
+// CHECK-SAME:  [[PORT:%.*]])
+// CHECK:       [[TMP0:%.*]] = tail call i32 asm sideeffect "inl ${1:k}, ${0:b}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 [[PORT]])
```
[the remark this answers is not in the document]
Comment (MalaySanghi) [reply clang/test/CodeGen/X86/ms-x86-intrinsics.c]: In llvm-IR this remain ax. 
There were other issues which I fixed. I had only tested the outp end-end and modelled this after the outp intrinsic. I've tested this end-end now.
```
_ind(port)
```
becomes
```
%0 = tail call i32 asm sideeffect "inl ${1:w}, ${0:k}", "={ax},N{dx},~{dirflag},~{fpsr},~{flags}"(i16 %port) #3, !srcloc !6
```
and eventually
```
inl	%dx, %eax
```

File: clang/lib/Headers/intrin.h [hunk 6]
```diff
@@ -329,6 +329,28 @@ static __inline__ void __DEFAULT_FN_ATTRS __stosq(unsigned __int64 *__dst,
...
+}
+
+static inline unsigned long _inpd(unsigned short port) {
+  unsigned long ret;
+  __asm__ volatile("inl %w1, %k0" : "=a"(ret) : "Nd"(port));
+  return ret;
+}
+
+#define inp(port) _inp((port))
+#define inpw(port) _inpw((port))
```
Comment (phoebewang) [inline clang/lib/Headers/intrin.h:352]: From MSVC doc:

> The inp and inpw names are older, deprecated names for the _inp and _inpw functions.

Maybe we don't need to provide them now.

Comment (MalaySanghi) [reply clang/lib/Headers/intrin.h:352]: I don't have a strong preference. I can remove them

File: clang/lib/Headers/intrin.h [hunk 7]
```diff
@@ -329,6 +329,25 @@ static __inline__ void __DEFAULT_FN_ATTRS __stosq(unsigned __int64 *__dst,
 static __inline__ void __DEFAULT_FN_ATTRS __halt(void) {
   __asm__ volatile("hlt");
 }
+
+static inline int _inp(unsigned short port) {
+  int ret;
+  __asm__ volatile("inb %w1, %b0" : "=a"(ret) : "Nd"(port));
```
Comment (phoebewang) [inline clang/lib/Headers/intrin.h:335]: I took a look at the Clang code, it seems these `b`, `w` modifiers are only meaningful to ARM/AArch64. We don't need them for X86 assembly.

Comment (phoebewang) [reply clang/lib/Headers/intrin.h:335]: I have a check, they are still required.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Review by RKSimon — COMMENTED** [review]
Maybe add this to ReleaseNotes?

**Pull request merged**